src/Juvix/Core/FromFrontend.hs

```diff
@@ -0,0 +1,260 @@
+  deriving
+    ( HasSource "param" (P.Parameterisation primTy primVal),
+      HasReader "param" (P.Parameterisation primTy primVal)
+    )
+    via ReaderField "param" (EnvAlias primTy primVal)
+  deriving
+    ( HasSource "core" (IR.Globals primTy primVal),
+      HasSink   "core" (IR.Globals primTy primVal),
+      HasState  "core" (IR.Globals primTy primVal)
+    )
+    via StateField "core" (EnvAlias primTy primVal)
+
+throwFF :: Error -> Env primTy primVal a
+throwFF = throw @"fromFrontendError"
+
+
+paramConstant :: P.Parameterisation primTy primVal
+              -> FE.Constant -> Maybe primVal
+paramConstant p (FE.Number (FE.Double'  d)) = P.floatVal  p d
+paramConstant p (FE.Number (FE.Integer' i)) = P.intVal    p i
+paramConstant p (FE.String (FE.Sho      s)) = P.stringVal p s
+
+
+transformTerm :: FE.Expression -> Env primTy primVal (IR.Term primTy primVal)
+transformTerm = fmap hrToIR . transformTermHR
+
+transformTermHR :: FE.Expression -> Env primTy primVal (HR.Term primTy primVal)
+transformTermHR (FE.Constant k) = do
+  p <- ask @"param"
+  case paramConstant p k of
+    Just x  -> pure $ HR.Prim x
+    Nothing -> throwFF $ UnsupportedConstant k
+transformTermHR (FE.Let l) = _
+transformTermHR (FE.LetType l) = _
+transformTermHR (FE.Match m) = _
+transformTermHR (FE.Name n) = pure $ HR.Elim $ HR.Var _n
+transformTermHR (FE.Lambda l) = _
+transformTermHR (FE.Application (FE.App f xs)) = do
+  f' <- toElim f =<< transformTermHR f
+  HR.Elim . foldl HR.App f' <$> traverse transformTermHR xs
+transformTermHR (FE.Primitive (FE.Prim p)) = do
+  param <- ask @"param"
+  maybe (throwFF $ UnknownPrimitive p) pure $
+    primTy param p <|> primVal param p
+ where
+  primTy  param p = HR.PrimTy <$> HM.lookup p (P.builtinTypes  param)
+  primVal param p = HR.Prim   <$> HM.lookup p (P.builtinValues param)
+transformTermHR (FE.List (FE.ListLit es)) =
+  makeList <$> traverse transformTermHR es
+transformTermHR (FE.Tuple (FE.TupleLit es)) =
+  makeTuple <$> traverse transformTermHR es
+transformTermHR (FE.Block (FE.Bloc e)) =
+  transformTermHR e
+transformTermHR (FE.ExpRecord (FE.ExpressionRecord fs)) =
+  makeRecord <$> traverse traverseField (toList fs)
+ where
+  traverseField (FE.NonPunned n e) = (n,) <$> transformTermHR e
+transformTermHR (FE.ArrowE a) = transformArrow a
+transformTermHR (FE.NamedTypeE (FE.NamedType' _ e)) =
+  -- TODO the name will only become relevant (outside of arrows)
+  -- when refinements are supported
+  transformTermHR e
+transformTermHR (FE.RefinedE r) =
+  throwFF $ RefinementsUnsupported r
+transformTermHR (FE.UniverseName i) =
+  -- TODO for universe polymorphism
+  throwFF $ UniversesUnsupported i
+transformTermHR (FE.Parened e) = transformTermHR e
+
+transformArrow :: FE.ArrowExp -> Env primTy primVal (HR.Term primTy primVal)
+transformArrow f@(FE.Arr' xa π b) =
+  case xa of
+    FE.NamedTypeE (FE.NamedType' x a) -> go π (getName x) a b
+    a                                 -> go π _unusedName a b
```

Comment:
is `_unusedName` a hole that gets fileld? should I use this to replace undefined that go unused?

Reply:
It's a GHC hole at the moment because I wasn't sure what to put there. I suppose just `""` would do, if that won't break anything.